src/front/wgsl/number.rs

```diff
@@ -0,0 +1,506 @@
+                    return None;
+                }
+
+                let number = general_extract.end(bytes);
+
+                let (bytes, kind) =
+                    consume_map!(bytes, [b'f' => FloatKind::F32, b'h' => FloatKind::F16]);
+
+                Some(ExtractInfo {
+                    info: Info::HexFloat { number, kind },
+                    rest: to_str(bytes),
+                })
+            } else {
+                Some(ExtractInfo {
+                    info: Info::HexFloatMissingExponent {
+                        significand,
+                        kind: None,
+                    },
+                    rest: to_str(bytes),
+                })
+            }
+        } else {
+            if consumed == 0 {
+                return None;
+            }
+
+            let significand = general_extract.end(bytes);
+            let digits = digits_extract.end(bytes);
+
+            let exp_extract = ExtractSubStr::start(bytes);
+
+            let (bytes, has_exp) = consume!(bytes, b'p' | b'P');
+
+            if has_exp {
+                let (bytes, _) = consume!(bytes, b'+' | b'-');
+                let (bytes, consumed) = consume_dec_digits(bytes);
+
+                if consumed == 0 {
+                    return None;
+                }
+
+                let exponent = exp_extract.end(bytes);
+
+                let (bytes, kind) =
+                    consume_map!(bytes, [b'f' => FloatKind::F32, b'h' => FloatKind::F16]);
+
+                Some(ExtractInfo {
+                    info: Info::HexFloatMissingPeriod {
+                        significand,
+                        exponent,
+                        kind,
+                    },
+                    rest: to_str(bytes),
+                })
+            } else {
+                let (bytes, kind) =
+                    consume_map!(bytes, [b'i' => IntKind::I32, b'u' => IntKind::U32]);
+
+                Some(ExtractInfo {
+                    info: Info::HexInt {
+                        is_negative,
+                        digits,
+                        kind,
+                    },
+                    rest: to_str(bytes),
+                })
+            }
+        }
+    } else {
+        let (_, is_first_zero) = consume!(bytes, b'0');
+
+        let (bytes, consumed) = consume_dec_digits(bytes);
+
+        let (bytes, has_dot) = consume!(bytes, b'.');
+
+        if has_dot {
+            let (bytes, consumed2) = consume_dec_digits(bytes);
+
+            if consumed + consumed2 == 0 {
+                return None;
+            }
+
+            let (bytes, has_exp) = consume!(bytes, b'e' | b'E');
+
+            let bytes = if has_exp {
+                let (bytes, _) = consume!(bytes, b'+' | b'-');
+                let (bytes, consumed) = consume_dec_digits(bytes);
+
+                if consumed == 0 {
+                    return None;
+                }
+
+                bytes
+            } else {
+                bytes
+            };
```

Comment:
My thought was that letting `consume!` update bytes as a side effect would make odd plumbing like this go away.